### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Romero10**
### Add User:

**Summary**: Response code is NOT 400 when creating a new user with the same name and sex as existing user in the system

**Preconditions**: I am authorized user.

**Steps**:
1.Send GET request to /zip-codes endpoint

**Actual Result**: Response code is 424
**Expected Result**: Response code should be 400.

**Automation Test Coverage**
|| Test Class || Test Name ||
| TC_UserAdd | verifyAddNotUniqueUserTest |

--------------------------------------------------------------------------------
### Delete User:

**Summary**: User with required fields is NOT removed

**Preconditions**: I am authorized user.

**Steps**:
1. Send DELETE request to /users endpoint where user with required fields

**Actual Result:** User is NOT removed.
**Expected Result:** User should be removed.

**Automation Test Coverage**
|| Test Class || Test Name ||
| TC_UserDelete | verifyRemoveUserWithRequiredFieldsTest|

--------------------------------------------------------------------------------
### Update User:

**Summary**: User is removed when updating a user with incorrect zip code

**Preconditions**: I am authorized user.

**Steps**:
1. Send PUT/PATCH request to /users endpoint to update a user  - where some new values and new zip code is incorrect

**Actual Result:** User is removed.
**Expected Result**: User should NOT be removed.

**Automation Test Coverage**
|| Test Class || Test Name ||
| TC_UserUpdate | verifyUpdateUserWithIncorrectZipCodeTest |

--------------------------------------------------------------------------------
### Upload User:

**Summary**: 500 Internal Server Error when uploading a few users with incorrect zip code.

**Preconditions**: I am authorized user.

**Steps**:
1. Send POST request to /users/upload endpoint with json file where array of users to upload and at least one user has incorrect (unavailable) zip code

Example file:

[
  {
    "age": "561",
    "name": "Us",
    "sex": "MALE",
    "zipCode": "12345"
  },
  {
    "age": "56",
    "name": "Us1",
    "sex": "MALE",
    "zipCode": "55555"
  }
]

**Actual Result**: 500 Internal Server Error. Added user with correct zip code.
**Expected Result**: Response code should be 424. Users should not be uploaded.

**Automation Test Coverage**
|| Test Class || Test Name ||
| TC_UserUpload | verifyUserUploadWithIncorrectZipCodeTest |

--------------------------------------------------------------------------------
**Summary**: 500 Internal Server Error when uploading a few users with required fields are missed.

**Preconditions**: I am authorized user.

**Steps:**
1. Send POST request to /users/upload endpoint with json file where array of users to upload and at least one user has missed required field

Example file:

[
  {
    "age": "561",
    "name": "Us",
    "sex": "MALE",
    "zipCode": "12345"
  },
  {
    "age": "56",
    "name": null,
    "sex": "MALE",
    "zipCode": "ABCDE"
  }
]

**Actual Result**: 500 Internal Server Error. Added user with correct required fields.
**Expected Result:** Response code should be 409. User should NOT be removed.

**Automation Test Coverage**
|| Test Class || Test Name ||
| TC_UserUpload | verifyUserUploadWithRequiredFieldsMissedTest |

### CODE REVIEW

File: src/main/java/api/training/dto/UserDto.java [hunk 1]
```diff
@@ -1,11 +1,13 @@
 package api.training.dto;
 
 import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.EqualsAndHashCode;
 import lombok.Getter;
 import lombok.Setter;
 
 @Getter
 @Setter
+@EqualsAndHashCode
```
Comment (EugenBorisik) [inline src/main/java/api/training/dto/UserDto.java]: This annotation takes all fields, but User has complex key by required fields only, please update

File: src/main/java/api/training/services/BaseService.java [hunk 2]
```diff
@@ -3,19 +3,22 @@
...
 import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.RequestBuilder;
 import org.apache.http.impl.client.BasicCredentialsProvider;
 import org.apache.http.message.BasicHeader;
 
 public class BaseService {
 
 	protected static Client client;
+	protected static ObjectMapper mapper;
```
Comment (EugenBorisik) [inline src/main/java/api/training/services/BaseService.java]: Can be final

File: src/main/java/api/training/services/BaseService.java [hunk 3]
```diff
@@ -25,13 +28,20 @@
...
 		client = new Client(provider);
+		mapper = new ObjectMapper();
 	}
 
 	protected static Header getTokenHeader(String scope) {
 		TokenDto token = AuthenticationService.getInstance().getToken(scope);
 		return new BasicHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token.getAccessToken());
 	}
 
+	protected static int getResponseStatusCode(RequestBuilder request) {
```
Comment (EugenBorisik) [inline src/main/java/api/training/services/BaseService.java]: Does it make sense to pass not RequestBuilder, but Request by itself?

File: src/main/java/api/training/services/ZipCodeService.java [hunk 4]
```diff
@@ -37,14 +37,15 @@ private ZipCodeService() {
...
-				.build();
-		CloseableHttpResponse response = client.request(request);
-		closeResponse();
-		return response.getStatusLine().getStatusCode();
+				.addHeader(HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType())
+				.setEntity(new StringEntity(zipCodes.toString(), ContentType.APPLICATION_JSON));
+		return getResponseStatusCode(request);
+	}
+
+	public static boolean isZipCodeExist() {
```
Comment (EugenBorisik) [inline src/main/java/api/training/services/ZipCodeService.java]: Probably it is better to rename into "isAnyZipCodeAvailable"

File: src/main/java/api/training/services/UserService.java [hunk 5]
```diff
@@ -21,31 +27,107 @@ private UserService() {
...
-		int statusCode = response.getStatusLine().getStatusCode();
-		List<UserDto> listOfZipCodes = Arrays.stream(client.parseResponseTo(UserDto[].class, response))
-				.collect(Collectors.toList());
-		closeResponse();
-		return Pair.of(statusCode, listOfZipCodes);
+	public static Pair<Integer, List<UserDto>> getUsers(Sex sex) {
+		RequestBuilder request = getUserRequest(SCOPE_READ, HttpGet.METHOD_NAME);
+		request.addParameter("sex", sex.getSexName());
+		return getUserResponse(request);
 	}
```
Comment (EugenBorisik) [inline src/main/java/api/training/services/UserService.java]: It is possible to get users by sex and age in one request. Could you please create method, which gets users by all possible parameters and other methods will be like:
`... getUsers(Sex sex) {
    return getUsers(sex, null);
}
`

File: src/main/java/api/training/dto/UserDto.java [hunk 6]
```diff
@@ -21,9 +23,29 @@
 	private String zipCode;
 
 	public String toString() {
```
Comment (EugenBorisik) [inline src/main/java/api/training/dto/UserDto.java:32]: You don't need such complex toString() method, use ObjectMapper to convert object into json string

File: src/main/java/api/training/dto/UpdateUserDto.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.Setter;
+
+@Setter
+@Getter
+public class UpdateUserDto {
+
+	private UserDto userNewValues;
+	private UserDto userToChange;
+
+	public String toString() {
```
Comment (EugenBorisik) [inline src/main/java/api/training/dto/UpdateUserDto.java:13]: You don't need such complex toString() method, use ObjectMapper to convert object into json string

File: src/main/java/api/training/services/UserService.java [hunk 8]
```diff
@@ -21,31 +27,107 @@ private UserService() {
...
+
+	public static List<UserDto> getUsersFromFile(String fileName) {
+		List<UserDto> userDto;
+		try {
+			File file = new File(System.getProperty("user.dir") + "/testData/" + fileName);
+			userDto = Arrays.stream(mapper.readValue(file, UserDto[].class)).collect(Collectors.toList());
+		} catch (IOException e) {
+			throw new Exceptions.JsonFileParseToUserModelException(e);
+		}
+		return userDto;
```
Comment (EugenBorisik) [inline src/main/java/api/training/services/UserService.java]: You can remove this line and put return in line 127

File: src/test/java/user/TC_UserAdd.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+			ZipCodeService.addZipCodes(Lists.newArrayList(zipCode));
+		}
+	}
+
+	@Test
+	public void verifyAddUserAndRemoveZipCodeTest() {
+		Pair<Integer, List<String>> availableZipCodes = ZipCodeService.getAvailableZipCodes();
+		String randomZipCode = availableZipCodes.second().get(RandomUtils.nextInt(0, availableZipCodes.second().size() - 1));
+
+		String uniqueName = NAME + RandomStringUtils.randomAlphabetic(4);
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserAdd.java]: Move this line in @BeforeMethod

File: src/test/java/user/TC_UserAdd.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+		softAssert.assertAll();
+	}
+
+	@Test
+	public void verifyAddUserWithIncorrectZipCodeTest() {
+		String uniqueName = NAME + RandomStringUtils.randomAlphabetic(4);
+		userDto.setName(uniqueName);
+		userDto.setAge(RandomUtils.nextInt(1, 99));
+		userDto.setSex(Sex.values()[RandomUtils.nextInt(0, Sex.values().length)]);
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserAdd.java]: I think you can create method in Sex enum: Sex.getRandom();

File: src/test/java/user/TC_UserAdd.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import org.testng.asserts.SoftAssert;
+import org.testng.internal.collections.Pair;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TC_UserAdd {
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserAdd.java:21]: Please add one more test: try to create user without required fields

Comment (EugenBorisik) [inline src/test/java/user/TC_UserAdd.java:21]: My recommendation about test structure: try to follow AAA-pattern. Prepare test data, make actions, assert. F.e. on my mind this test is more readable:
```
		String uniqueName = NAME + RandomStringUtils.randomAlphabetic(4);
		userDto.setName(uniqueName);
		userDto.setSex(Sex.values()[RandomUtils.nextInt(0, Sex.values().length)]);
		userDtoList.add(userDto);

		int statusCode = UserService.createUser(userDto);

		softAssert.assertEquals(statusCode, HttpStatus.SC_CREATED,
				"Response code is NOT 201 when creating a new user with required fields.");
		softAssert.assertEquals(UserService.findUsersByName(uniqueName).size(),
				1, "User with required fields is NOT added to application.");

		softAssert.assertAll();
```

File: src/test/java/user/TC_UserUpdate.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+		UpdateUserDto updateUser = new UpdateUserDto();
+		updateUser.setUserToChange(userDto);
+		updateUser.setUserNewValues(userUpdateDto);
+
+		int statusCode = UserService.updateUser(updateUser);
+		userDtoList.add(userUpdateDto);
+		softAssert.assertEquals(statusCode, HttpStatus.SC_OK,
+				"Response code is NOT 200 when updating a user.");
+
+		int actualAge = UserService.findUsersByName(userUpdateDto.getName()).get(0).getAge();
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserUpdate.java]: 1. What if application have 2 users with the same name, but different sex? You need to use complex key here.
2. Please add assertion, that actual list of users doesn't have old userDto, only updated user.

File: src/test/java/user/TC_UserUpdate.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+		softAssert.assertEquals(actualAge, newAge, "User is NOT updated.");
+		softAssert.assertAll();
+	}
+
+	@Test
+	public void verifyUpdateUserWithIncorrectZipCodeTest() {
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserUpdate.java]: Add assertion, that initial user wasn't deleted from application and its zip code is not in the available list

File: src/test/java/user/TC_UserUpdate.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+		softAssert.assertFalse(users.isEmpty(), "Updated user does not exist in users.");
+
+		softAssert.assertAll();
+
+		int actualAge = users.get(0).getAge();
+		Assert.assertEquals(actualAge, newAge, "User is updated when updating a user with incorrect zip code.");
+	}
+
+	@Test
+	public void verifyUpdateUserWithRequiredFieldsMissedTest() {
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserUpdate.java]: This test verifies update operation for userNewValues without required fields. Please add one more test when userToChange doesn't have required fields

File: src/test/java/user/TC_UserUpdate.java [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import org.testng.asserts.SoftAssert;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TC_UserUpdate {
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserUpdate.java:22]: Why are you using only put method?

File: src/test/java/user/TC_UserDelete.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+		Pair<Integer, List<String>> availableZipCodes = ZipCodeService.getAvailableZipCodes();
+		softAssert.assertTrue(availableZipCodes.second().contains(zipCode),
+				"Zip code is NOT returned in list of available zip codes when removing a user with required fields.");
+
+		softAssert.assertAll();
+	}
+
+	@Test
+	public void verifyRemoveUserAnyRequiredFieldMissedTest() {
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserDelete.java:99]: Let's add assertion that zip code is not returned in available list

File: src/test/java/user/TC_UserUpload.java [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import org.testng.asserts.SoftAssert;
+import org.testng.internal.collections.Pair;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TC_UserUpload {
```
Comment (EugenBorisik) [inline src/test/java/user/TC_UserUpload.java:20]: Potentially this class must have all tests from "create users" test class

Comment (EugenBorisik) [inline src/test/java/user/TC_UserUpload.java:20]: Add test, when file is invalid (incorrect json structure f.e.)

**Review by EugenBorisik — CHANGES_REQUESTED** [review]

File: src/main/java/api/training/services/BaseService.java (hunk 2 above)
Comment (Romero10) [reply src/main/java/api/training/services/BaseService.java]: Added.

File: src/main/java/api/training/services/ZipCodeService.java (hunk 4 above)
Comment (Romero10) [reply src/main/java/api/training/services/ZipCodeService.java]: Renamed.

File: src/main/java/api/training/dto/UserDto.java (hunk 6 above)
Comment (Romero10) [reply src/main/java/api/training/dto/UserDto.java:32]: Fixed.

File: src/main/java/api/training/dto/UpdateUserDto.java (hunk 7 above)
Comment (Romero10) [reply src/main/java/api/training/dto/UpdateUserDto.java:13]: Fixed.

File: src/main/java/api/training/services/UserService.java (hunk 5 above)
Comment (Romero10) [reply src/main/java/api/training/services/UserService.java]: Created one more method using all parameters.

File: src/main/java/api/training/services/UserService.java (hunk 8 above)
Comment (Romero10) [reply src/main/java/api/training/services/UserService.java]: Fixed.

File: src/test/java/user/TC_UserAdd.java (hunk 9 above)
Comment (Romero10) [reply src/test/java/user/TC_UserAdd.java]: Moved the line 51 in @BeforeMethod.

File: src/test/java/user/TC_UserAdd.java (hunk 10 above)
Comment (Romero10) [reply src/test/java/user/TC_UserAdd.java]: Fixed.

File: src/main/java/api/training/services/BaseService.java (hunk 3 above)
Comment (Romero10) [reply src/main/java/api/training/services/BaseService.java]: This method was removed during refactoring.

File: src/test/java/user/TC_UserDelete.java (hunk 16 above)
Comment (Romero10) [reply src/test/java/user/TC_UserDelete.java:99]: Added one more assertion.

File: src/test/java/user/TC_UserAdd.java (hunk 11 above)
Comment (Romero10) [reply src/test/java/user/TC_UserAdd.java:21]: Added one more test.

File: src/test/java/user/TC_UserUpdate.java (hunk 13 above)
Comment (Romero10) [reply src/test/java/user/TC_UserUpdate.java]: Added.

File: src/test/java/user/TC_UserUpdate.java (hunk 14 above)
Comment (Romero10) [reply src/test/java/user/TC_UserUpdate.java]: Added one more test where 'userToChange' doesn't have required fields.

File: src/test/java/user/TC_UserUpdate.java (hunk 15 above)
Comment (Romero10) [reply src/test/java/user/TC_UserUpdate.java:22]: Fixed.

File: src/test/java/user/TC_UserUpdate.java (hunk 12 above)
Comment (Romero10) [reply src/test/java/user/TC_UserUpdate.java]: Fixed.

File: src/test/java/user/TC_UserUpload.java (hunk 17 above)
Comment (Romero10) [reply src/test/java/user/TC_UserUpload.java:20]: Added.

**Review by EugenBorisik — APPROVED** [review]

**Pull request merged**